cladecombiner/aggregator.py

```diff
@@ -0,0 +1,278 @@
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from copy import copy
+from dataclasses import dataclass
+from warnings import warn
+
+from .ordered_taxon import OrderedTaxon
+from .taxon import Taxon
+from .taxonomy_scheme import PhylogeneticTaxonomyScheme
+from .utils import table
+
+
+@dataclass
+class TaxonMapping:
+    """
+    Class for tracking proposed step in aggregation process.
+    """
+
+    map: dict[Taxon, Taxon]
+    "from : to dictionary of the mapping"
+    done: dict[Taxon, bool]
+    "to : add_to_stack dictionary tracking if the aggregated taxon should be added back to the stack"
+
+
+class Aggregator(ABC):
+    """
+    A partially-abstract base class for aggregation.
+
+    Descendants need only implement next_agg_step() to create a usable aggregation object.
+    """
+
+    def __init__(
+        self,
+        in_taxa: Iterable[Taxon],
+    ):
+        """
+        Constructor for aggregation infrastructure shared by all subclasses.
+
+        Parameters
+        ----------
+        in_taxa : Iterable[Taxon]
+            The taxa we wish to aggregate.
+        """
+        self.input_taxa = list(in_taxa)
+        "The taxa we wish to aggregate."
+        self.stack = copy(self.input_taxa)
+        "Taxa that still need to be aggregated."
+        self.messy_map = {}
```

Comment:
Why do we need to record this history? Is it for debugging?

I would imagine that in general aggregation is a Markov process?

Reply:
I don't follow about Markov processes?

We definitely don't need it for what we've got here. But in this step-by-step case, it's easier for me to conceive of recording all the steps and collating them into the final map later in one sweep, than constantly editing/amending the not-quite-final map until it's done.

Reply:
By "Markov," I meant "memoryless." My mental picture was that each aggregation step takes in the current aggregation state, and maybe some other information (eg the tree structure), and outputs the next state of the aggregation. Where aggregation state is the map {input_taxon => output_taxon}